Drop WTF_ALLOW_UNSAFE_BUFFER_USAGE use in _WKTouchEventGenerator.mm & WKContentViewInteraction.mm

cdumez opened the pull request:
#### e2d50f6d2ed3b5a93aab1a4889104f8285ac9ac0
<pre>
Drop WTF_ALLOW_UNSAFE_BUFFER_USAGE use in _WKTouchEventGenerator.mm &amp; WKContentViewInteraction.mm
<a href="https://bugs.webkit.org/show_bug.cgi?id=282412">https://bugs.webkit.org/show_bug.cgi?id=282412</a>

Reviewed by Darin Adler.

* Source/WebKit/UIProcess/_WKTouchEventGenerator.mm:
(-[_WKTouchEventGenerator _updateTouchPoints:window:]):
(-[_WKTouchEventGenerator touchDownAtPoints:touchCount:window:]):
(-[_WKTouchEventGenerator liftUpAtPoints:touchCount:window:]):
(-[_WKTouchEventGenerator moveToPoints:touchCount:duration:window:]):
(-[_WKTouchEventGenerator _updateTouchPoints:count:window:]): Deleted.
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _touchStartDeferringGestures]):
(-[WKContentView _touchEndDeferringGestures]):

Canonical link: <a href="https://commits.webkit.org/286021@main">https://commits.webkit.org/286021@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1f3dacaf0ddf0ae6cd3543753f20ee7a515a4a18

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/74514 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/53943 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/27325 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/78914 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/25752 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/63076 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/1728 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/58560 "Found 88 new test failures: fast/css-custom-paint/out-of-memory-while-adding-worklet-module.html fast/css/focus-ring-exists-for-search-field.html fast/images/object-data-url-case-insensitivity.html fast/inline/list-marker-inside-container-with-margin.html fast/inline/min-content-width-with-hypens.html fast/inline/overflowing-content-with-hypens.html fast/mediastream/canvas-video-to-canvas.html fast/mediastream/captureStream/canvas3d.html fast/multicol/table-vertical-align.html fast/ruby/annotation-with-line-gap.html ... (failure)") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16863 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/77581 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48719 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/64070 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38962 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/45783 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/24085 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/67126 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21912 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/80422 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/1831 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/1082 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66833 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/1979 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/64088 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66118 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10062 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/8214 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11501 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/1796 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/4583 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/1824 "Built successfully") | | | 
| | [❌ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/1813 "Hash 1f3dacaf for PR 36021 does not build (failure)") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/1831 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebKit/UIProcess/_WKTouchEventGenerator.mm:
```diff
@@ -48,7 +47,7 @@
 NSUInteger const HIDMaxTouchCount = 5;
 
-static int fingerIdentifiers[HIDMaxTouchCount] = { 2, 3, 4, 5, 1 };
+static constexpr std::array<int, 5> fingerIdentifiers { 2, 3, 4, 5, 1 };
```
Thanks for switching to `constexpr`.

Does not seem like an improvement to move from `HIDMaxTouchCount` to `5` here. If we are specifying a size, I think it’s great to document that must match `HIDMaxTouchCount`. Or if we instead want to omit the size then, since the values are all `int` constants, this can just be `std::array`, with no template arguments.

@aproskuryakov has also pointed out to me a few times that in cases like this we only need `constexpr`, don’t also need `static`.

cdumez replied:
Oh, that's a mistake, I intended to use `HIDMaxTouchCount` instead of 5 here, my bad.

darinadler commented on Source/WebKit/UIProcess/_WKTouchEventGenerator.mm:
```diff
@@ -249,38 +248,39 @@ - (BOOL)_sendMarkerHIDEventInWindow:(UIWindow *)window completionBlock:(void (^)
 
     // Update point locations.
-    for (NSUInteger i = 0; i < count; ++i)
+    for (size_t i = 0; i < points.size(); ++i)
         _activePoints[i].point = points[i];
```
I think @weinig showed me a slightly more elegant style for iterating a pair of arrays in a case like this with a modern for loop without having to use an index, maybe by zippering the two arrays together? Would love to use that if it’s elegant and memorable enough. Comes up 3 or 4 times in this pull request.

cdumez replied:
I found:
```
std::array container1 = { 1, 2, 3, 4, 5  };
std::array container2 = { 6, 7, 8, 9, 10 };

for(auto& [a, b] : std::views::zip(container1, container2))
    // Do something.
```

but I think that requires C++23 (https://en.cppreference.com/w/cpp/ranges/zip_view). I'm not familiar with a nicer way to do this in C++20.

darinadler replied:
If we can’t use this yet in WebKit due to what is available in the libraries and compilers, I understand. Once we can use it, I think that style will be much better than indexing, much harder to get wrong.

darinadler replied:
We can also add a simple `zip` to WTF and use it while waiting to move our minimum C++ library version forward. My research indicates it’s straightforward, so doesn’t really have to be in the standard library.

geoffreygaren replied:
Not quite as nice as zip in this use case, but I recently added:

```
for (auto [ i, point ] : IndexedRange(points)) {
    _activePoints[i].point = point;
}
```

cdumez replied:
Ok, following up in https://github.com/WebKit/WebKit/pull/36063

darinadler commented on Source/WebKit/UIProcess/_WKTouchEventGenerator.mm:
```diff
@@ -249,38 +248,39 @@ - (BOOL)_sendMarkerHIDEventInWindow:(UIWindow *)window completionBlock:(void (^)
+- (void)touchDownAtPoints:(CGPoint*)rawLocations touchCount:(NSUInteger)touchCount window:(UIWindow *)window
 {
     touchCount = std::min(touchCount, HIDMaxTouchCount);
+    auto locations = unsafeMakeSpan(rawLocations, touchCount);
```
I kind of feel like we want to tighten this up when we can, with a top level method or function that **only** does the conversion to `span`.

When do we use _forge_ vs. _make_ as the word in the function name for creating a `span` from a raw pointer?

cdumez replied:
unsafeForgeSpan was renamed to unsafeMakeSpan yesterday :)

darinadler replied:
For the record, I’m happy with that rename!

darinadler commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -2332,7 +2330,7 @@ - (void)_inspectorNodeSearchRecognized:(UIGestureRecognizer *)gestureRecognizer
 
     // 1) Sort the points horizontally.
-    WebCore::FloatPoint points[4] = { quad.p1(), quad.p4(), quad.p2(), quad.p3() };
+    std::array<WebCore::FloatPoint, 4> points { quad.p1(), quad.p4(), quad.p2(), quad.p3() };
```
Can this just be `std::array` without the template arguments?

darinadler commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -2402,7 +2400,7 @@ - (WKTouchEventsGestureRecognizer *)touchEventGestureRecognizer
 - (NSArray<WKDeferringGestureRecognizer *> *)_touchStartDeferringGestures
 {
-    WKDeferringGestureRecognizer *recognizers[3];
+    std::array<WKDeferringGestureRecognizer *, 3> recognizers;
```
While it’s slightly more efficient to use `std::array`, it seems like it would be nicer to use something where the counting and conversion to `NSArray` is more built in rather than hand coded. For example we could use a `Vector` with inline capacity and add only a tiny bit of overhead; would not add any memory allocation. And would be more obviously safe, even if the magic number `3` was wrong.

darinadler replied:
And then we’d just use `makeNSArray` instead of calling the `NSArray` class method directly with multiple arguments, relieving us of the need to check that the two arguments are consistent.

cdumez replied:
Looks like I could also keep using a std::array and just do:
```
return createNSArray(std::span { recognizers }.first(count)).autorelease();
```

cdumez replied:
Although now I feel like it may simply be better to construct a NSMutableArray right away instead of requiring an intermediary `std::array` / `Vector`

cdumez replied:
I think this is nice and efficient:
```
static void appendRecognizerIfNonNull(RetainPtr<NSMutableArray>& array, const RetainPtr<WKDeferringGestureRecognizer>& recognizer)
{
    if (recognizer)
        [array addObject:recognizer.get()];
}

- (NSArray<WKDeferringGestureRecognizer *> *)_touchStartDeferringGestures
{
    RetainPtr recognizers = adoptNS([[NSMutableArray alloc] initWithCapacity:3]);
    appendRecognizerIfNonNull(recognizers, _touchStartDeferringGestureRecognizerForImmediatelyResettableGestures);
    appendRecognizerIfNonNull(recognizers, _touchStartDeferringGestureRecognizerForDelayedResettableGestures);
    appendRecognizerIfNonNull(recognizers, _touchStartDeferringGestureRecognizerForSyntheticTapGestures);
    return recognizers.autorelease();
}
```

darinadler reviewed (approved):
Everything is just fine as is, but I do have some ideas for refinement either now or in future.

cdumez committed:
Drop WTF_ALLOW_UNSAFE_BUFFER_USAGE use in _WKTouchEventGenerator.mm & WKContentViewInteraction.mm
https://bugs.webkit.org/show_bug.cgi?id=282412

Reviewed by Darin Adler.

* Source/WebKit/UIProcess/_WKTouchEventGenerator.mm:
(-[_WKTouchEventGenerator _updateTouchPoints:window:]):
(-[_WKTouchEventGenerator touchDownAtPoints:touchCount:window:]):
(-[_WKTouchEventGenerator liftUpAtPoints:touchCount:window:]):
(-[_WKTouchEventGenerator moveToPoints:touchCount:duration:window:]):
(-[_WKTouchEventGenerator _updateTouchPoints:count:window:]): Deleted.
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _touchStartDeferringGestures]):
(-[WKContentView _touchEndDeferringGestures]):

Canonical link: https://commits.webkit.org/286021@main
